### ISSUE TITLE
Charging won't start

### ISSUE THREAD

**Issue by Ralle001** [sebdehne/dehneevse-mqtt-bridge#2]
Car says: Low voltage in system

When I start the MQTT App it always reconnects after 10 secs, I presume this is intended.
![](https://i.imgur.com/3FhRNil.png)

After I connect the cable to the car I get a ton of "notify" messages:
![](https://i.imgur.com/XqjDCgo.png)

And it doesn't matter when I send the message with the charging rate and the contactor on it always ends with the error.

Also when I send the data collect message it raises and error.
Im gonna attach a few pics below:
![](https://i.imgur.com/S7W2gAf.png)

![](https://i.imgur.com/oceV736.png)

I would greatly appreciate your help.

**Comment by Ralle001** [comment]
The problem is with the 109th line in the evse_handler.rs. The payload_buf is empty for some reason. @sebdehne can you please help?

**Comment by Ralle001** [comment]
I managed to fix the "REQUEST_DATA_COLLECTION". Change this line (evse_handler.rs):
```
let mut payload_buf: Vec<u8> = Vec::with_capacity(msg_length as usize);
```
To this:
```
let mut payload_buf: Vec<u8> = Vec::new();
payload_buf.resize(msg_length as usize, 0);
```

**Comment by Ralle001** [comment]
But still not working. It looks like it starts charging for a half second but then it drops it. Also the output of the data collection looks kinda dodgy:
```
2023 -09-01T14: 29: 51Z INFO  dehneevse_mqtt_bridge: :mqtt_handler
]MQTT: publishing msg from EVSE: {
"message_type": "request_data_collection",
"client_id": "Charger 3",
"pwm_percent": 15,
"contactor_state": true,
"measurements": {
    "pilot_voltage": "volt_9",
    "proximity_pilot_amps": "no_cable",
    "phase1_millivolts": 3139,
    "phase2_millivolts": 41,
    "phase3_millivolts": 3100,
    "phase1_milliamps": 41,
    "phase2_milliamps": 35,
    "phase3_milliamps": 83,
    "wifi_rssi": -60,
    "uptime_milliseconds": 437837,
    "current_control_pilot_adc": 2443,
    "current_proximity_pilot_adc": 3638,
    "logging_buffer": "421212: Pilot voltage changed from 0 to 1. ADC: 2459\u0000421214: Notify dataChanged sent\u0000421841: Pilot voltage changed from 1 to 0. ADC: 2631\u0000421842: Notify dataChanged sent\u0000421860: Pilot voltage changed from 0 to 1. ADC: 2353\u0000421861: Notify dataChanged sent\u0000422506: Pilot voltage changed from 1 to 0. ADC: 2635\u0000422508: Notify dataChanged sent\u0000422525: Pilot voltage changed from 0 to 1. ADC: 2397\u0000422526: Notify dataChanged sent\u0000423154: Pilot voltage changed from 1 to 0. ADC: 2732\u0000423155: Notify dataChanged sent\u0000423172: Pilot voltage changed from 0 to 1. ADC: 2351\u0000423173: Notify dataChanged sent\u0000423347: Pilot voltage changed from 1 to 0. ADC: 2665\u0000423349: Notify dataChanged sent\u0000436781: Pilot voltage changed from 0 to 1. ADC: 2443\u0000436783: Notify dataChanged sent\u0000"
}
}
```
The cable should be connected
@sebdehne

**Comment by Ralle001** [comment]
The arduino communicates with the Contactor, right? @sebdehne 
Because when I plug the cable in the car the contactor doesn't turn on. Only if I send the message through the MQTT

**Comment by Ralle001** [comment]
I measured the voltage between the CP and PE and when I turn on and off the contactor it shows the correct amount but when I send a message with the updated Amp percentage then it changes to a minus amount, for example -12 Volt which is correspondent to Error. @sebdehne

**Comment by sebdehne** [comment]
The CP signal is a PWM signal and the voltage is constantly changing between -12 and +12. So then I wonder how you measure the volage? Are you using an oscilloscope or a multimeter?

Sorry for not being very responsive, on this topic, but I do not really have the time.

**Comment by Ralle001** [comment]
> The CP signal is a PWM signal and the voltage is constantly changing between -12 and +12. So then I wonder how you measure the volage? Are you using an oscilloscope or a multimeter?
> 
> 
> 
> Sorry for not being very responsive, on this topic, but I do not really have the time.

I use a multimeter for checking the voltage. But for me it doesn't change, it is constantly a subzero number. Also there is another problem, the arduino says it has no cables when I plug in the charging cord, but the car seems to trying to start charging.

**Comment by Ralle001** [comment]
@sebdehne 
So I hardwired the 13Amp cable to the firmware but still the car detects the connection with the charger but when I send the charging rate and turn on the contactor nothing happens. It seems like it wanna charge but probably there is not enough Amps in the cable. Did you face this problem ever?

**Comment by Ralle001** [comment]
> @sebdehne So I hardwired the 13Amp cable to the firmware but still the car detects the connection with the charger but when I send the charging rate and turn on the contactor nothing happens. It seems like it wanna charge but probably there is not enough Amps in the cable. Did you face this problem ever?

So apparently if I start the arduino program from my PC using PlatformIO it works, it tracks when the cable connects but if it is NOT connected to the computer it just doesn't track if the cable is connected.

**Comment by Ralle001** [comment]
I just bought a oscilloscope and I tested it on a original charger to see the PWM signal. It works perfectly with the genuine charger but for the DehneEVSE one there is no PWN signal.

**Comment by sebdehne** [comment]
Yes, the square wave is only started when the EVSE offers the car to start charging. For that to happen, you need enable charging by sending a "request_set_pwm_percent" command.

To figure out in what state you are, there are a few things you can check. First, what voltage do you read on the CP? What happens when you plug in the charging cable? You should see a change in voltage.

You can also compile the firmware with [DEBUG](https://github.com/sebdehne/DehneEVSE-Firmware/blob/master/include/config.h#L6) enabled and connect a USB cable and monitor the serial output. You should get some logging.

**Comment by Ralle001** [comment]
> Yes, the square wave is only started when the EVSE offers the car to start charging. For that to happen, you need enable charging by sending a "request_set_pwm_percent" command.
> 
> To figure out in what state you are, there are a few things you can check. First, what voltage do you read on the CP? What happens when you plug in the charging cable? You should see a change in voltage.
> 
> You can also compile the firmware with [DEBUG](https://github.com/sebdehne/DehneEVSE-Firmware/blob/master/include/config.h#L6) enabled and connect a USB cable and monitor the serial output. You should get some logging.

I monitored the data and as previously mentioned I have a problem with the voltage read. It always reads a way higher data (circa 3300mV) so it will say Volt_12 and no_cable.

**Comment by Ralle001** [comment]
I went for another testing session and I was able to find the PWM, sort of. I am still missing the 9V signal, when I set the percentage I get the 12V signal but when I connect the charger to the car it immediately drops to 6V, there is no in between. Still thew problem might be that the charger doesn't detect the cable. In that regard when I turned on the DEBUG setting that you mentioned I saw the without the cable the CP and PP is around 2500mV but when I connect the cable it doesn't drop, it goes up to 3000mV.

**Comment by Ralle001** [comment]
> I went for another testing session and I was able to find the PWM, sort of. I am still missing the 9V signal, when I set the percentage I get the 12V signal but when I connect the charger to the car it immediately drops to 6V, there is no in between. Still thew problem might be that the charger doesn't detect the cable. In that regard when I turned on the DEBUG setting that you mentioned I saw the without the cable the CP and PP is around 2500mV but when I connect the cable it doesn't drop, it goes up to 3000mV.

@sebdehne Do you have any idea? Im kinda stuck here

**Comment by Ralle001** [comment]
Still the same. I guess the CP mV is used for the mode (volt_9, volt_6), I was able to see that it is working. But for the no_cable issue it is not solved. The PP value is always around 3000mV.

Also one more thing that I noticed, when I request data from the charger through the MQTT software the phase1_millivolts is under 100, while the phase2_millivolts and phase3_millivolts are around 3000.
@sebdehne
